cpp/src/arrow/csv/parser.h

```diff
@@ -0,0 +1,136 @@
+// under the License.
+
+#ifndef ARROW_CSV_PARSER_H
+#define ARROW_CSV_PARSER_H
+
+#include <cstdint>
+#include <vector>
+
+#include "arrow/csv/options.h"
+#include "arrow/status.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+namespace csv {
+
+constexpr int32_t kMaxParserNumRows = 100000;
+
+// Whether BlockParser will use bitfields for better memory consumption
+// and cache locality.
+#undef CSV_PARSER_USE_BITFIELD
+#define CSV_PARSER_USE_BITFIELD
+
+/// \class BlockParser
+/// \brief A reusable block-based parser for CSV data
+///
+/// The parser takes a block of CSV data and delimits rows and fields,
+/// unquoting and unescaping them on the fly.  Parsed data is own by the
+/// parser, so the original buffer can be discarded after Parse() returns.
+///
+/// If the block is truncated (i.e. not all data can be parsed), it is up
+/// to the caller to arrange the next block to start with the trailing data.
+/// Also, if the previous block ends with CR (0x0d) and a new block starts
+/// with LF (0x0a), the parser will consider the leading newline as an empty
+/// line; the caller should therefore strip it.
+class ARROW_EXPORT BlockParser {
+ public:
+  explicit BlockParser(ParseOptions options, int32_t num_cols = -1,
+                       int32_t max_num_rows = kMaxParserNumRows);
+
+  /// \brief Parse a block of data
+  ///
+  /// Parse a block of CSV data, ingesting up to max_num_rows rows.
+  /// The number of bytes actually parsed is returned in out_size.
+  Status Parse(const char* data, uint32_t size, uint32_t* out_size);
+
+  /// \brief Parse the final block of data
+  ///
+  /// Like Parse(), but called with the final block in a file.
+  /// The last row may lack a trailing line separator.
+  Status ParseFinal(const char* data, uint32_t size, uint32_t* out_size);
+
+  int32_t num_rows() const { return num_rows_; }
+  int32_t num_cols() const { return num_cols_; }
+  /// \brief Return the total size in bytes of parsed data
+  int32_t num_bytes() const { return static_cast<int32_t>(parsed_.size()); }
+
+  /// \brief Visit parsed values in a column
+  ///
+  /// The signature of the visitor is (const uint8_t* data, uint32_t size, bool quoted)
+  template <typename Visitor>
+  Status VisitColumn(int32_t col_index, Visitor&& visit) const {
+#ifdef CSV_PARSER_USE_BITFIELD
+    for (int32_t pos = col_index; pos < num_cols_ * num_rows_; pos += num_cols_) {
+      auto start = values_[pos].offset;
+      auto stop = values_[pos + 1].offset;
+      auto quoted = values_[pos + 1].quoted;
+      ARROW_RETURN_NOT_OK(visit(parsed_.data() + start, stop - start, quoted));
+    }
+#else
+    for (int32_t pos = col_index; pos < num_cols_ * num_rows_; pos += num_cols_) {
+      auto start = offsets_[pos];
+      auto stop = offsets_[pos + 1];
+      auto quoted = quoted_[pos];
+      ARROW_RETURN_NOT_OK(visit(parsed_.data() + start, stop - start, quoted));
+    }
+#endif
+    return Status::OK();
```

Comment:
So, the way that things are structured here is similar to the way that the pandas CSV reader works. There are some issues with this:

* The conversion hot path involves strided memory access
* Conversions to Arrow varbinary involves an extra conversion

After this patch is merged, I would like to experiment with a "columnar" tokenization where we actually construct the Arrow varbinary layout on a column-by-column basis as we are tokenizing the data, so the following occurs:

* The type conversion hot path does not do strided memory access; conversion performance should be better
* If the column's type is string, no further copies are necessary

I don't know for sure, but I expect this to be a performance win. I can dig into this in the coming weeks

Reply:
Hmm, I don't think we can avoid a strided memory read in any case. We are converting from a row-wise layout (CSV) to a column-wise layout (Arrow). Doing the strided memory read here is not worse than doing it somewhere else.

The varbinary conversion is another thing. I agree that if we get a CSV that is mostly binary / string columns, we might make parsing a bit faster still by avoiding an extra copy here. My intuition is that most useful CSV files (in the Arrow context) will have numeric columns whose parsing performance is slow enough that micro-optimizing binary columns isn't really important.

Reply:
> Doing the strided memory read here is not worse than doing it somewhere else.

What I'm actually saying is not to do strided memory access at all. You can build contiguous buffers for each column while you are tokenizing. 

I'll spend some time working on this and flamegraphing / etc. so that I can assure myself of the best approach to tokenize-convert. I have been unsatisfied for years about the pandas CSV reader but never motivated enough to do anything about it because of pandas's underlying internal problems. I'd like to turn this CSV reader into something that we can use for the next 20 years (!)

Reply:
> You can build contiguous buffers for each column while you are tokenizing.

That's true. But then you're doing strided memory writes, which I'm unsure is better.